File: nodejs/src/updater.ts
/**
 * 程式內置更新模組
 *
 * 功能：
 *   - 查詢當前 / 遠端最新版本 (git commit hash)
 *   - 執行 git pull 更新程式碼
 *   - 自動重啟進程
 */

import { spawn, execSync, type SpawnOptions } from "node:child_process";
import { utimesSync } from "node:fs";
import { closeDb } from "./db/database.js";

// ========================
// 型別定義
// ========================

export interface VersionInfo {
  /** 短 commit hash，例如 "abc1234" */
  hash: string;
  /** ISO 8601 提交時間，例如 "2024-01-15T10:30:00+08:00" */
  date: string;
  /** 提交訊息第一行 */
  message: string;
}

export interface UpdateCheckResult {
  /** 是否有更新 */
  hasUpdate: boolean;
  /** 當前版本 */
  current: VersionInfo;
  /** 最新版本 */
  latest: VersionInfo;
  /** 落後的 commit 數量 */
  commitsBehind: number;
  /** 落後的 commit 列表（每行一條） */
  newCommits: string[];
}

export interface UpdateResult {
  success: boolean;
  message: string;
  /** 更新後的 commit hash */
  newHash?: string;
}

// ========================
// Git 輔助函數
// ========================

function execGit(args: string[]): string {
  try {
    const result = execSync(`git ${args.join(" ")}`, {
      cwd: process.cwd(),
      encoding: "utf-8",
      timeout: 30_000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    return result.trim();
  } catch (err: any) {
    const stderr = err.stderr?.toString()?.trim() ?? "";
    const stdout = err.stdout?.toString()?.trim() ?? "";
    throw new Error(stderr || stdout || err.message || "git command failed");
  }
}

function parseVersionInfo(ref: string): VersionInfo {
  const hash = execGit(["rev-parse", "--short", ref]);
  const date = execGit(["log", "-1", "--format=%cI", ref]);
  const message = execGit(["log", "-1", "--format=%s", ref]);
  return { hash, date, message };
}

// ========================
// 公開 API
// ========================

/**
 * 取得當前版本資訊
 */
export function getCurrentVersion(): VersionInfo {
  return parseVersionInfo("HEAD");
}

/**
 * Fetch 遠端並取得最新版本資訊
 * 呼叫後 origin/main 會更新到最新
 */
export function fetchAndCheckUpdate(): UpdateCheckResult {
  // Step 1: Fetch latest from remote
  execGit(["fetch", "origin", "main"]);

  const current = parseVersionInfo("HEAD");
  const latest = parseVersionInfo("origin/main");

  // Step 2: Compare hashes
  const currentFull = execGit(["rev-parse", "HEAD"]);
  const latestFull = execGit(["rev-parse", "origin/main"]);
  const hasUpdate = currentFull !== latestFull;

  // Step 3: Get commits behind (if any)
  let newCommits: string[] = [];
  let commitsBehind = 0;

  if (hasUpdate) {
    const logOutput = execGit([
      "log", "--oneline", "--no-decorate",
      `HEAD..origin/main`,
    ]);
    if (logOutput) {
      newCommits = logOutput.split("\n").filter(Boolean);
      commitsBehind = newCommits.length;
    }
  }

  return { hasUpdate, current, latest, commitsBehind, newCommits };
}

/**
 * 檢查工作目錄是否乾淨（沒有未提交的更改）
 */
export function isWorkingDirClean(): boolean {
  const status = execGit(["status", "--porcelain"]);
  return status.length === 0;
}

/**
 * 執行 git pull 更新程式碼
 */
export function performUpdate(): UpdateResult {
  try {
    // Check working directory
    if (!isWorkingDirClean()) {
      return {
        success: false,
        message: "工作目錄有未提交的更改，請先處理後再更新。",
      };
    }

    // Pull latest code
    execGit(["pull", "origin", "main"]);

    // Get new version
    const newHash = execGit(["rev-parse", "--short", "HEAD"]);

    return {
      success: true,
      message: `更新成功！新版本：${newHash}`,
      newHash,
    };
  } catch (err: any) {
    return {
      success: false,
      message: `更新失敗：${err.message}`,
    };
  }
}

/**
 * 重啟進程
 *
 * 分兩種模式處理：
 *   1. tsx watch 模式：觸發 entry file 的 mtime 變化，tsx watcher 會自動重啟
 *   2. 生產模式（node dist/...）：spawn 新進程（detached），然後 exit 舊進程
 *
 * 延遲時間確保 Telegram 訊息能先送達。
 */
export function restartProcess(delayMs = 2000): void {
  console.log(`[updater] 將在 ${delayMs}ms 後重啟...`);

  setTimeout(() => {
    const argvStr = process.argv.slice(1).join(" ");
    const isWatch =
      (argvStr.includes("tsx") && argvStr.includes("watch")) ||
      process.env.TSX_WATCH === "true";

    // 關閉資料庫（兩種模式都需要）
    try {
      closeDb();
    } catch (e) {
      console.error("[updater] 關閉資料庫失敗：", e);
    }

    if (isWatch) {
      // tsx watch 模式：觸發 file change 讓 watcher 自動重啟
      console.log("[updater] tsx watch 模式：觸發 watcher 重啟...");
      try {
        const entryFile = process.argv[process.argv.length - 1];
        const now = new Date();
        utimesSync(entryFile, now, now);
      } catch (e) {
        console.error("[updater] 觸發 watcher 重啟失敗，直接退出：", e);
      }
      console.log("[updater] 舊進程正在退出...");
      process.exit(0);
      return;
    }

    // 生產模式：spawn 新的 detached 進程
    console.log("[updater] 正在啟動新進程...");
    const child = spawn(
      process.execPath,
      process.argv.slice(1),
      {
        detached: true,
        stdio: "inherit",
        cwd: process.cwd(),
        env: process.env,
      } satisfies SpawnOptions,
    );
    child.unref();

    console.log("[updater] 舊進程正在退出...");
    process.exit(0);
  }, delayMs);
}

/**
 * 更新並重啟（git pull + restart）
 */
export function updateAndRestart(): UpdateResult {
  const result = performUpdate();
  if (result.success) {
    restartProcess(2000);
  }
  return result;
}

File: python/updater.py
"""
程式內置更新模組

功能：
  - 查詢當前 / 遠端最新版本 (git commit hash)
  - 執行 git pull 更新程式碼
  - 自動重啟進程 (os.execv)
"""

from __future__ import annotations

import asyncio
import logging
import os
import subprocess
import sys
from dataclasses import dataclass

logger = logging.getLogger(__name__)


# ============================================================
# 資料結構
# ============================================================


@dataclass
class VersionInfo:
    """版本資訊"""
    hash: str           # 短 commit hash
    date: str           # ISO 8601 提交時間
    message: str        # 提交訊息第一行


@dataclass
class UpdateCheckResult:
    """更新檢查結果"""
    has_update: bool
    current: VersionInfo
    latest: VersionInfo
    commits_behind: int
    new_commits: list[str]


@dataclass
class UpdateResult:
    """更新執行結果"""
    success: bool
    message: str
    new_hash: str | None = None


# ============================================================
# Git 輔助函數
# ============================================================


def _run_git(args: list[str], timeout: int = 30) -> str:
    """執行 git 命令，回傳 stdout（已 strip）"""
    try:
        result = subprocess.run(
            ["git"] + args,
            capture_output=True,
            text=True,
            timeout=timeout,
            check=False,
        )
        if result.returncode != 0:
            err = (result.stderr or "").strip()
            out = (result.stdout or "").strip()
            raise RuntimeError(err or out or f"git {' '.join(args)} failed (exit {result.returncode})")
        return (result.stdout or "").strip()
    except subprocess.TimeoutExpired:
        raise RuntimeError(f"git {' '.join(args)} timed out after {timeout}s")


def _parse_version_info(ref: str) -> VersionInfo:
    """從 git ref 取得版本資訊"""
    hash_ = _run_git(["rev-parse", "--short", ref])
    date = _run_git(["log", "-1", "--format=%cI", ref])
    message = _run_git(["log", "-1", "--format=%s", ref])
    return VersionInfo(hash=hash_, date=date, message=message)


# ============================================================
# 公開 API
# ============================================================


def get_current_version() -> VersionInfo:
    """取得當前版本資訊"""
    return _parse_version_info("HEAD")


def fetch_and_check_update() -> UpdateCheckResult:
    """
    Fetch 遠端並檢查是否有更新。
    呼叫後 origin/main 會更新到最新。
    """
    # Step 1: Fetch latest from remote
    _run_git(["fetch", "origin", "main"])

    current = _parse_version_info("HEAD")
    latest = _parse_version_info("origin/main")

    # Step 2: Compare hashes
    current_full = _run_git(["rev-parse", "HEAD"])
    latest_full = _run_git(["rev-parse", "origin/main"])
    has_update = current_full != latest_full

    # Step 3: Get commits behind (if any)
    new_commits: list[str] = []
    commits_behind = 0

    if has_update:
        log_output = _run_git([
            "log", "--oneline", "--no-decorate",
            "HEAD..origin/main",
        ])
        if log_output:
            new_commits = [line for line in log_output.split("\n") if line]
            commits_behind = len(new_commits)

    return UpdateCheckResult(
        has_update=has_update,
        current=current,
        latest=latest,
        commits_behind=commits_behind,
        new_commits=new_commits,
    )


def is_working_dir_clean() -> bool:
    """檢查工作目錄是否乾淨（沒有未提交的更改）"""
    status = _run_git(["status", "--porcelain"])
    return len(status) == 0


def perform_update() -> UpdateResult:
    """執行 git pull 更新程式碼"""
    try:
        if not is_working_dir_clean():
            return UpdateResult(
                success=False,
                message="工作目錄有未提交的更改，請先處理後再更新。",
            )

        _run_git(["pull", "origin", "main"])

        new_hash = _run_git(["rev-parse", "--short", "HEAD"])

        return UpdateResult(
            success=True,
            message=f"更新成功！新版本：{new_hash}",
            new_hash=new_hash,
        )
    except Exception as e:
        return UpdateResult(
            success=False,
            message=f"更新失敗：{e}",
        )


async def restart_process(delay: float = 2.0) -> None:
    """
    重啟進程（非阻塞方式）。

    使用 asyncio.create_task 排程延遲重啟，
    讓呼叫端可以先回覆 Telegram 訊息。

    重啟方式：os.execv 直接替換當前進程。
    """
    logger.info("[updater] 將在 %.1f 秒後重啟...", delay)

    async def _do_restart():
        await asyncio.sleep(delay)

        # 刷新使用量佇列，避免資料遺失
        try:
            from db import database
            await database._flush_usage_queue()
        except Exception as e:
            logger.error("[updater] 刷新使用量佇列失敗：%s", e)

        logger.info("[updater] 正在重啟進程...")
        # os.execv 替換當前進程 — 乾淨的重啟
        # 新進程從 main() 開始，重新初始化一切
        os.execv(sys.executable, [sys.executable] + sys.argv)

    asyncio.create_task(_do_restart())
